src/main/java/software/amazon/smithy/lsp/ext/SmithyProject.java

```diff
@@ -161,35 +160,97 @@ public File getRoot() {
-            if (shape.getType() != ShapeType.MEMBER) {
-                if (locations.containsKey(shapeName)) {
-                    locations.get(shapeName).add(location);
+
+            for (Shape shape : shapes) {
+                SourceLocation sourceLocation = shape.getSourceLocation();
+                Position startPosition = new Position(sourceLocation.getLine() - 1, sourceLocation.getColumn() - 1);
+                Position endPosition;
+                if (endMarker < sourceLocation.getLine()) {
+                    endPosition = new Position(sourceLocation.getLine() - 1, sourceLocation.getColumn() - 1);
                 } else {
-                    List<Location> locList = new ArrayList<Location>();
-                    locList.add(location);
-                    locations.put(shapeName, locList);
+                    endPosition = new Position(endMarker, 0);
                 }
-            }
-        });
 
+                // If a shape is not a member, move the end marker for setting the next shape location.
+                if (shape.getType() != ShapeType.MEMBER) {
+                    endMarker = startPosition.getLine();
+                    List<Trait> traits = new ArrayList<>(shape.getAllTraits().values());
+                    // If the shape has traits, advance the end marker again.
+                    if (!traits.isEmpty()) {
+                        traits.sort(new TraitSorter());
+                        endMarker = traits.get(0).getSourceLocation().getLine() - 1;
+                    }
+                    // Move the end marker when encountering line comments or empty lines.
+                    if (lines.size() > endMarker) {
+                        while (lines.get(endMarker - 1).trim().startsWith("//")
+                                || lines.get(endMarker - 1).trim().equals("")) {
+                            endMarker = endMarker - 1;
+                        }
+                    }
+                }
+                Location location =  new Location(getUri(modelFile), new Range(startPosition, endPosition));
+                addLocationToMap(shape, locations, location);
+            }
+        }
         return locations;
     }
 
+    private static List<String> getFileLines(Path path) {
+        try {
+            if (Utils.isSmithyJarFile(path.toString())) {
+                return Utils.jarFileContents(path.toString());
+            }
+            return Files.readAllLines(path);
+        } catch (IOException e) {
+            LspLog.println("Path " + path + " was found in temporary buffer");
+        }
+        return Collections.emptyList();
+    }
+
+    private static void addLocationToMap(Shape shape, Map<String, List<Location>> locations, Location location) {
+        String shapeName = shape.getId().getName();
+        // Members get the same shapeName as their parent structure
+        // so we ignore them, to avoid producing a location per-member
+        // TODO: index members somehow as well?
```

Comment:
This seems useful, since `apply` syntax allows for targeting a member externally. Maybe the `location` map should use the `ShapeId` as the key instead of the name?

Reply:
When finding a token in a line, SmithyTextDocumentService only returns a potential shape name. We could use the shape name to try to find the appropriate `ShapeId` and then change over the location map to use the `ShapeId` as the key.

Reply:
When finding a token in a line, SmithyTextDocumentService only returns a potential shape name. We could use the shape name to try to find the appropriate `ShapeId` and then change over the location map to use the `ShapeId` as the key.